nameko/messaging.py

```diff
@@ -130,71 +158,31 @@ def serializer(self):
             SERIALIZER_CONFIG_KEY, DEFAULT_SERIALIZER
         )
 
-    @property
-    def retry(self):
-        """ Enable automatic retries when publishing a message that fails due
-        to a connection error.
-
-        Retries according to :attr:`self.retry_policy`.
-        """
-        return True
-
-    @property
-    def retry_policy(self):
-        """ Policy to apply when retrying message publishes, if enabled.
-
-        See :attr:`self.retry`.
-        """
-        return DEFAULT_RETRY_POLICY
-
     def setup(self):
 
-        exchange = self.exchange
-        queue = self.queue
-
         verify_amqp_uri(self.amqp_uri)
 
         with get_connection(self.amqp_uri) as conn:
-            if queue is not None:
-                maybe_declare(queue, conn)
-            elif exchange is not None:
-                maybe_declare(exchange, conn)
+            for entity in self.declare:
+                maybe_declare(entity, conn)
+
+        serializer = self.options.pop('serializer', self.serializer)
+
+        self.publisher = self.Publisher(
+            self.amqp_uri,
+            serializer=serializer,
+            exchange=self.exchange,
+            declare=self.declare,
+            **self.options
+        )
 
     def get_dependency(self, worker_ctx):
-        def publish(msg, **kwargs):
-            exchange = self.exchange
-            serializer = self.serializer
-
-            if exchange is None and self.queue is not None:
-                exchange = self.queue.exchange
-
-            retry = kwargs.pop('retry', self.retry)
-            retry_policy = kwargs.pop('retry_policy', self.retry_policy)
-            mandatory = kwargs.pop('mandatory', False)
-
-            with get_producer(self.amqp_uri, self.use_confirms) as producer:
-                headers = self.get_message_headers(worker_ctx)
-                producer.publish(
-                    msg, exchange=exchange, headers=headers,
-                    serializer=serializer, retry=retry,
-                    retry_policy=retry_policy, mandatory=mandatory,
-                    **kwargs
-                )
+        propagate_headers = self.get_message_headers(worker_ctx)
```

Comment:
how come this var is called `propagate_` and not `extra_`? (propagate sounds like a bool to me)